File: evals/eval_research_agent.py
"""Research Agent evaluation focused on web search and information retrieval."""

import os
import re
import sys
from pathlib import Path
from typing import Any

# Ensure project root is on sys.path
project_root = Path(__file__).resolve().parents[1]
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

from autoevals import LLMClassifier  # noqa: E402
from braintrust import Eval  # noqa: E402
from dotenv import load_dotenv  # noqa: E402

from evals.parameters import ResearchAgentPromptParam, ResearchModelParam  # noqa: E402
from src.agents.research_agent import get_research_agent  # noqa: E402
from src.helpers import run_adk_agent  # noqa: E402
from src.tracing import configure_adk_tracing  # noqa: E402

load_dotenv()
configure_adk_tracing(
    api_key=os.environ.get("BRAINTRUST_API_KEY"),
    project_id=os.environ.get("BRAINTRUST_PROJECT_ID"),
    project_name=os.environ.get("BRAINTRUST_PROJECT", "vercel-ai-sdk-supervisor"),
)


def _param_value(param: Any, default: Any) -> Any:
    if param is None:
        return default
    if hasattr(param, "value"):
        return getattr(param, "value")
    if isinstance(param, type):
        try:
            instance = param()
            if hasattr(instance, "value"):
                return getattr(instance, "value")
        except Exception:
            return default
    return param


async def run_research_task(input: dict, hooks: Any = None) -> dict:
    """Run a research query through the research agent."""
    try:
        params = hooks.parameters if hooks and hasattr(hooks, "parameters") else {}
        research_agent_prompt = _param_value(params.get("research_agent_prompt"), None)
        research_model = _param_value(params.get("research_model"), "gemini-2.0-flash-lite")

        agent = get_research_agent(
            system_prompt=research_agent_prompt,
            model=research_model,
        )
        query = str(input.get("query", ""))

        run_result = await run_adk_agent(
            agent=agent,
            query=query,
            app_name="google-adk-supervisor-eval-research",
        )
        serialized = run_result["messages"]

        tool_calls: list[str] = []
        for msg in serialized:
            for tc in msg.get("tool_calls", []):
                tool_calls.append(str(tc.get("name", "")))

        if hooks and hasattr(hooks, "metadata"):
            hooks.metadata.update(
                {
                    "tool_calls": tool_calls,
                    "used_web_search": "tavily_search" in tool_calls,
                    "total_messages": len(serialized),
                }
            )

        return {"messages": serialized}
    except Exception as e:
        if hooks and hasattr(hooks, "metadata"):
            hooks.metadata.update({"error": str(e)})
        return {"messages": [{"error": str(e)}]}


RESEARCH_TEST_DATA = [
    {
        "input": {"query": "Who is the current president of France?"},
        "expected": {"should_use_search": True, "should_have_url": True},
    },
    {
        "input": {"query": "What is the capital of Japan?"},
        "expected": {"should_use_search": True, "should_have_url": True},
    },
    {
        "input": {"query": "When was the Eiffel Tower built?"},
        "expected": {"should_use_search": True, "should_have_url": True},
    },
    {
        "input": {"query": "What are the main causes of climate change?"},
        "expected": {"should_use_search": True, "should_have_url": True},
    },
]


async def web_search_usage_scorer(output, metadata=None):
    """Check if the agent used web search when appropriate."""
    if metadata and metadata.get("used_web_search"):
        return 1.0
    return 0.0


async def source_attribution_scorer(output):
    """Check if the response includes URL citations."""
    messages = output.get("messages", [])
    for msg in reversed(messages):
        content = msg.get("content", "") if isinstance(msg, dict) else ""
        role = msg.get("role", "") if isinstance(msg, dict) else ""
        if content and role == "assistant":
            if re.search(r"https?://", content):
                return 1.0
            break
    return 0.0


async def efficiency_scorer(output, metadata=None):
    """Score based on minimal tool calls."""
    if not metadata:
        return 0.5

    num_searches = metadata.get("tool_calls", []).count("tavily_search")
    if num_searches == 1:
        return 1.0
    if num_searches == 2:
        return 0.9
    if num_searches <= 4:
        return 0.7
    return 0.5


answer_quality_prompt = """
You are evaluating a research agent's response to a factual question.

Question: {{input}}
Response: {{output}}

Evaluate the response on:
1. ACCURACY: Is the information correct and factual?
2. COMPLETENESS: Does it answer the question fully?
3. CLARITY: Is it well-structured and clear?
4. RELEVANCE: Does it address what was asked?

Respond with:
EXCELLENT - Accurate, complete, clear, and highly relevant
GOOD - Mostly accurate and complete with minor issues
FAIR - Some accuracy or completeness issues
POOR - Inaccurate, incomplete, or irrelevant
"""

answer_quality_scorer = LLMClassifier(
    name="Answer Quality",
    prompt_template=answer_quality_prompt,
    choice_scores={"EXCELLENT": 1.0, "GOOD": 0.75, "FAIR": 0.5, "POOR": 0.0},
    use_cot=True,
    model=os.environ.get("EVAL_JUDGE_MODEL", "gpt-4o"),
)


Eval(
    os.environ.get("BRAINTRUST_PROJECT", "vercel-ai-sdk-supervisor"),
    experiment_name="research-agent",
    data=RESEARCH_TEST_DATA,  # type: ignore
    task=run_research_task,
    scores=[
        web_search_usage_scorer,
        source_attribution_scorer,
        efficiency_scorer,
        answer_quality_scorer,
    ],  # type: ignore
    parameters={
        "research_agent_prompt": ResearchAgentPromptParam,
        "research_model": ResearchModelParam,
    },
)

File: evals/eval_research_agent.ts
import "dotenv/config";

import { Eval } from "braintrust";
import { z } from "zod";

import { DEFAULT_RESEARCH_AGENT_PROMPT } from "../src-ts/config.js";
import { runResearchAgent } from "../src-ts/agents/research-agent.js";
import { configureTracing } from "../src-ts/tracing.js";
import type { SerializedMessage } from "../src-ts/types.js";

configureTracing({
  apiKey: process.env.BRAINTRUST_API_KEY,
  projectName: process.env.BRAINTRUST_PROJECT ?? "vercel-ai-sdk-supervisor",
  projectId: process.env.BRAINTRUST_PROJECT_ID,
});

const parameters = {
  research_agent_prompt: z.string().default(DEFAULT_RESEARCH_AGENT_PROMPT),
  research_model: z.string().default("gpt-4.1-mini"),
};

type ResearchTaskInput = { query: string };
type ResearchTaskOutput = { messages: SerializedMessage[] };

async function runResearchTask(input: ResearchTaskInput, hooks: { parameters: { research_agent_prompt: string; research_model: string }; metadata: Record<string, unknown> }): Promise<ResearchTaskOutput> {
  try {
    const result = await runResearchAgent({
      query: input.query,
      systemPrompt: hooks.parameters.research_agent_prompt,
      model: hooks.parameters.research_model,
    });

    const toolCalls = result.messages.flatMap((message) => message.tool_calls ?? []);
    hooks.metadata.tool_calls = toolCalls;
    hooks.metadata.used_web_search = toolCalls.some((toolCall) => toolCall.name === "tavily_search");
    hooks.metadata.total_messages = result.messages.length;

    return { messages: result.messages };
  } catch (error) {
    hooks.metadata.error = String(error);
    return { messages: [{ role: "system", content: `error: ${String(error)}` }] };
  }
}

const RESEARCH_TEST_DATA = [
  { input: { query: "Who is the current president of France?" }, metadata: {} },
  { input: { query: "What is the capital of Japan?" }, metadata: {} },
  { input: { query: "When was the Eiffel Tower built?" }, metadata: {} },
  { input: { query: "What are the main causes of climate change?" }, metadata: {} },
];

function latestAssistant(messages: SerializedMessage[]): string {
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    const message = messages[index];
    if (message.role === "assistant" && message.content.trim()) {
      return message.content;
    }
  }
  return "";
}

async function webSearchUsageScorer(args: { output: ResearchTaskOutput }) {
  const toolNames = args.output.messages.flatMap((message) => message.tool_calls ?? []).map((toolCall) => toolCall.name);
  return { name: "Web Search Usage", score: toolNames.includes("tavily_search") ? 1 : 0 };
}

async function sourceAttributionScorer(args: { output: ResearchTaskOutput }) {
  const answerText = latestAssistant(args.output.messages);
  return { name: "Source Attribution", score: /https?:\/\//.test(answerText) ? 1 : 0 };
}

async function efficiencyScorer(args: { output: ResearchTaskOutput }) {
  const numSearches = args.output.messages
    .flatMap((message) => message.tool_calls ?? [])
    .filter((toolCall) => toolCall.name === "tavily_search").length;

  if (numSearches === 1) {
    return { name: "Efficiency", score: 1 };
  }
  if (numSearches === 2) {
    return { name: "Efficiency", score: 0.9 };
  }
  if (numSearches <= 4) {
    return { name: "Efficiency", score: 0.7 };
  }
  return { name: "Efficiency", score: 0.5 };
}

async function answerQualityScorer(args: { output: ResearchTaskOutput }) {
  const answerText = latestAssistant(args.output.messages);
  const hasText = answerText.trim().length > 0;
  const hasSource = /https?:\/\//.test(answerText);
  const score = hasText ? (hasSource ? 1 : 0.7) : 0;
  return { name: "Answer Quality", score };
}

await Eval(process.env.BRAINTRUST_PROJECT ?? "vercel-ai-sdk-supervisor", {
  experimentName: "research-agent",
  data: RESEARCH_TEST_DATA,
  task: runResearchTask,
  scores: [webSearchUsageScorer, sourceAttributionScorer, efficiencyScorer, answerQualityScorer],
  parameters,
});
